File: removeKdigits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Integer> sk = new Stack();
        StringBuilder sb = new StringBuilder();
        int index = 0;
        int count = k;

        if(num.length() == k){
            return "0";
        }

        while(index <= num.length()-1){
            if(sk.isEmpty()){
                sk.push(Integer.parseInt(String.valueOf(num.charAt(index))));
            }
            else if(!sk.isEmpty()){
                while(count > 0 && !sk.isEmpty() && sk.peek() > Integer.parseInt(String.valueOf(num.charAt(index)))){
                    sk.pop();
                    count--;
                }
                sk.push( Integer.parseInt(String.valueOf(num.charAt(index))));
            }
            index++;
        }
        while(count > 0){
            sk.pop();
            count--;
        }

        for(int val : sk){
            sb.append(val);
        }
        while(sb.length() > 1 && sb.charAt(0) == '0'){
            sb.deleteCharAt(0);
        }
        return sb.toString();
    }
}

File: removeKdigits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        count = k
        for i in num:
            if stack and stack[-1] > i:
                while stack and stack[-1] > i and count > 0:
                    stack.pop()
                    count-= 1
            stack.append(i)
        
        while count > 0:
            stack.pop()
            count -= 1

        i = 0
        while(i < len(stack) and stack[i] == "0"):
            i += 1
            
        return ''.join(stack[i:]) if (len(stack[i:]) > 0) else "0"
                
        
        